### ISSUE TITLE
视频生成质量

### ISSUE THREAD

**Issue by suuyyu98** [XuanhaoLiu/EEG2Video#39]
作者你好，在复现的时候，我生成得到的很多视频几乎是纯色块的，质量很低，请问得到的图像都是和论文一样高质量的吗？

**Comment by XuanhaoLiu** [comment]
你好，感谢你对我们工作的关注！
> 很多视频几乎是纯色块的，质量很低

这是正常现象，因为这个任务本质非常难，相当于脑电做40分类，而有些脑电任务4分类效果都差强人意。并且每一类样本量仅35，导致训练模型很容易过拟合到其中2-3类。
> 请问得到的图像都是和论文一样高质量的吗?

不是的，在一次生成中，对于测试集200个视频，只用少部分视频质量稍高，其余大部分都是质量低的视频。但是在进行定量结果分析时，不需要对其进行处理，直接将您生成的200段视频和GT的200段视频跑评测指标即可。

希望我的回答对你的工作有帮助，祝好！

**Comment by suuyyu98** [comment]
<img width="855" height="426" alt="Image" src="https://github.com/user-attachments/assets/4d0f7047-6225-43a4-8b80-8aed8efbaf37" />作者你好，这些定量的指标是和代码一样是3个正确样本吗？比如2-way是3个正确样本和一个随机的其他样本，模型判断是哪一个，所以随机的基线是0.75🤔在这个目录（EEG2Video/40_class_run_metrics.py）的代码选取的是3个正确样本，2-way的基线是3/4，40-way的基线是3/42
def n_way_top_k_acc(pred, class_id, n_way, num_trials=40, top_k=1):
    if isinstance(class_id, int):
        class_id = [class_id]
    pick_range =[i for i in np.arange(len(pred)) if i not in class_id]
    corrects = 0
    for t in range(num_trials):
        idxs_picked = np.random.choice(pick_range, n_way-1, replace=False)
        for gt_id in class_id:
            pred_picked = torch.cat([pred[gt_id].unsqueeze(0), pred[idxs_picked]])
            pred_picked = pred_picked.argsort(descending=False)[-top_k:]
            if 0 in pred_picked:
                corrects += 1
                break
    return corrects / num_trials, np.sqrt(corrects / num_trials * (1 - corrects / num_trials) / num_trials)

gt_class_id = model(**gt.to(device, torch.float16)).logits.argsort(-1,descending=False).detach().flatten()[-3:]   这里有点问题
        pred_out = model(**pred.to(device, torch.float16)).logits.softmax(-1).detach().flatten()

        acc, std = n_way_top_k_acc(pred_out, gt_class_id, n_way, num_trials, top_k)

**Comment by XuanhaoLiu** [comment]
感谢你仔细的对于指标的研究！事实上，我也没有仔细看过这份代码，而是直接使用了Mindvideo的评测指标：https://github.com/jqin4749/MindVideo/blob/master/src/MindVideo/utils/eval_metrics.py。

不过我看完了代码后，还是认为2-way top-1随机概率应该十分接近50%，可能不是严格的50%，但绝对没有到75%。
理想情况，视频/图片 分类器的top-1应该是GT truth，但我们都知道比如ImageNet中会有很多很细的分类，比如狗的子类有吉娃娃，拉布拉多等等，那么可能对于一张狗的图片，gt_class_id可能是dog，但是pred_logits最大的类别却是拉布拉多。导致实际上生成对了，但metric判断生成错了。

所以为了防止这种情况，我们取视频/图片 分类器的top-3，对于这3类，循环判断n-way top-1指标。但每次都是将n-1个错误的类和其中一个gt_class_id拼接：idxs_picked = np.random.choice(pick_range, n_way-1, replace=False)。

因此，对于一个1000分类的imagenet分类器，pick_range是997长度，idxs_picked始终是n-1长度。对于top-3的gt类别，每一个都与idxs_picked拼接，然后看pred_out中的最大值是否是拼接上的gt_class_id，因此每一次的随机水平仍然是1 / n，只是重复了3次(每次选取不同的类别作为gt_class_id)。然后整体重复了num_trial次。
